lib/src/modules/apply.cc

```diff
@@ -0,0 +1,251 @@
+# leave face require in for now until we for sure rule out pluginsync in process
+require 'puppet/face/plugin'
+require 'securerandom'
+require 'tempfile'
+# shell out for plugin download
+require 'open3'
+
+## TODO: Option to read from a file is for debugging. Only read from stdin in production.
+args = JSON.parse(ARGV[0] ? File.read(ARGV[0]) : STDIN.read)
+
+# Create temporary directories for all core Puppet settings so we don't clobber
+# existing state or read from puppet.conf. Also create a temporary modulepath.
+# Additionally include rundir, which gets its own initialization.
+puppet_root = Dir.mktmpdir
+moduledir = File.join(puppet_root, 'modules')
+Dir.mkdir(moduledir)
+cli = (Puppet::Settings::REQUIRED_APP_SETTINGS + [:rundir]).flat_map do |setting|
+  ["--#{setting}", File.join(puppet_root, setting.to_s.chomp('dir'))]
+end
+cli << '--modulepath' << moduledir
+Puppet.initialize_settings(cli)
+
+# Avoid extraneous output
+Puppet[:report] = false
+
+# Make sure to apply the catalog, TODO (on orch side): pipe through _noop
+Puppet[:noop] = args['_noop'] || false
+# TODO(on orch side): enable apply setting, currently this is only configurable from bolt.yaml in bolt
+if args['apply_settings']
+  args['apply_settings'].each do |setting, value|
+    Puppet[setting.to_sym] = value
+  end
+end
+
+Puppet[:default_file_terminus] = :file_server
+
+exit_code = 0
+begin
+  # This happens implicitly when running the Configurer, but we make it explicit here. It creates the
+  # directories we configured earlier.
+  Puppet.settings.use(:main)
+
+  # FAILED attempt. Being in an isolated environment breaks this approach
+  # We are not quite giving up on this, so i'm leaving it commented. For now we will move forward with
+  # sub-process approach to continue working on the feaure, but we should try again to do it in process
+  # if nothing else but to ensure total isolation.
+  # Do plugin sync with plugin face
+  #remote_env_for_plugins = Puppet::Node::Environment.remote(args['environment'])
+  #pool = Puppet::Network::HTTP::Pool.new(Puppet[:http_keepalive_timeout])
+  #Puppet[:plugindest] = args['plugin_cache']
+  #Puppet[:pluginfactdest] = args['plugin_cache']
+  #Puppet.override(:http_pool => pool) do
+  #  begin
+  #    handler = Puppet::Configurer::PluginHandler.new()
+  #    handler.download_plugins(remote_env_for_plugins)
+  #  ensure
+  #    pool.close
+  #  end
+  #end
+
+  plugin_cache = args['plugin_cache']
+  remote_env = args['environment']
+  begin
+    # shell out and let puppet do this operation as if it is in a "real" environment, this should be safe bc we are controlling the output dirs
+    # TODO: make invocation cross platform https://github.com/puppetlabs/pxp-agent/blob/579b883a5545f9a0895571d5f8cde329be5981ab/modules/pxp-module-puppet#L377-L384
+    stdout_str, stderr_str, status = Open3.capture3("/opt/puppetlabs/bin/puppet plugin download --plugindest #{plugin_cache} --pluginfactdest #{plugin_cache} --environment #{remote_env}")
```

Comment:
Setting `--plugindir` and `--pluginfactdir` is a bad bug. They need to be separate so the file resource does not conflict causing exactly *not* the caching behavior we want.